src/tests/functional/page_refresh_tests.js

```diff
@@ -180,12 +180,18 @@ test("uses morphing to update remote frames marked with refresh='morph'", async
 test("don't refresh frames contained in [data-turbo-permanent] elements", async ({ page }) => {
   await page.goto("/src/tests/fixtures/page_refresh.html")
 
+  await page.evaluate(() => {
+    const frame = document.getElementById("remote-permanent-frame");
+    if (frame) {
+      frame.textContent = "Frame to be preserved"
+    }
+  }) // The final assertion cannot be noNextEventOnTarget because the assertion passes even when it reloads the frame.
+
   await page.click("#form-submit")
   await nextEventNamed(page, "turbo:render", { renderMethod: "morph" })
   await nextBeat()
 
-  // Only the frame marked with refresh="morph" uses morphing
-  expect(await noNextEventOnTarget(page, "refresh-reload", "turbo:before-frame-morph")).toBeTruthy()
```

Comment:
This assertion was not working. When I commented out code and convinced myself the permanent frame WAS reloading, this test would still pass. I think it's a race condition where the assertion breezes through too quickly.

I tried multiple different strategies to make this test durable and the solution you see here is the only one that worked.